crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,337 @@
+use std::marker::PhantomData;
+
+use bevy_utils::{tracing::warn, HashSet};
+
+use crate::{component::ComponentId, prelude::Bundle, world::World};
+
+/// A rule about which [`Component`](crate::component::Component)s can coexist on entities
+///
+/// These rules must be true at all times for all entities in the [`World`].
+/// The generic [`Bundle`] type `B1` is always used in the `predicate`,
+/// while `B2` is used in the `consequence`.
+/// If only a single generic is provided, these types are the same.
+///
+/// When added to the [`World`], archetype invariants behave like [`assert!`];
+/// all archetype invariants must be true for every entity in the [`World`].
+/// Archetype invariants are checked each time [`Archetypes`](crate::archetype::Archetypes) is modified;
+/// this can occur on component addition, component removal, and entity spawning.
+///
+/// Archetypes are only modified when a novel archetype (set of components) is seen for the first time;
+/// swapping between existing archetypes will not trigger these checks.
+///
+/// Note that this is converted to an [`UntypedArchetypeInvariant`] when added to a [`World`].
+/// This is to ensure compatibility between different invariants.
+#[derive(Clone, Debug, PartialEq)]
+pub struct ArchetypeInvariant<B1: Bundle, B2: Bundle = B1> {
+    /// For all entities where the predicate is true
+    pub predicate: ArchetypeStatement<B1>,
```

Comment:
I think that predicate is reasonably clear, and is the standard term for boolean functions in CS. I don't feel too strongly though.

Reply:
I also don't have strong opinions there, but since the archetype invariant is essentially a logical implication, I thought borrowing the terminology from there would be more correct.

Reply:
I think that's fair then. Name changed to `premise` in 0dbc7d9e6c79630ad446deaffa08b6d956fdd738